src/librustc/ty/instance.rs

```diff
@@ -62,10 +62,39 @@ pub enum InstanceDef<'tcx> {
 }
 
 impl<'tcx> Instance<'tcx> {
+    /// Returns the `Ty` corresponding to this `Instance`,
+    /// with generic substitutions applied and lifetimes erased.
+    ///
+    /// This method can only be called when the 'substs' for this Instance
+    /// are fully monomorphic (no `ty::Param`'s are present).
+    /// This is usually the case (e.g. during codegen).
+    /// However, during constant evaluation, we may want
+    /// to try to resolve a `Instance` using generic parameters
+    /// (e.g. when we are attempting to to do const-propagation).
+    /// In this case, `Instace.ty_env` should be used to provide
+    /// the `ParamEnv` for our generic context.
     pub fn ty(&self, tcx: TyCtxt<'tcx>) -> Ty<'tcx> {
```

Comment:
Is this function used so much that we can't just add a param env arg to it? It seems too easy to get it wrong. If we want to keep two separate methods, we could rename this one to `monomorphic_ty`

Reply:
I don't think caller should have to to pass `ParamEnv::reveal_all` when they don't have a suitable `ParamEnv` to provide. I'll rename this method as you suggested.